### PR TITLE
Multithreading Interface Documentation

### LINKED ISSUE #62: Multithreading interface documentation
Create a README in the threading folder that details how to inherit and use the new threading interface.

### PR DESCRIPTION
**Pull request by ClayJay3**
### Threading Docs
Added a detailed guide on how developers should use the `AutonomyThread` interface. It goes over how to actually inherit the class and implement the pure virtual functions, as well as some basics on how resources should be handled with mutexes, locks, and atomics.

README is in the `src/threads` folder.

A thorough readthrough from a few different people would be nice to catch any errors/inaccuracies or make any suggestions about the formatting.

### CODE REVIEW

**Comment by Byrdman32** [comment]
Added Jesse as a reviewer since the Drone Autonomy Code will be based off of Autonomy_Software - Sprint 0

File: src/threads/README.md [hunk 1]
```diff
@@ -1,38 +1,204 @@
...
+
+**NOTE: Look in the `examples/threading` directory to see example implementations of what this guide discusses.**
+
+### Step 1: Create the class and inherit the interface.
+Create a new file in the `threads` directory and name it appropriately. For example, if there is a main class responsible for managing the states of the rover, create a file named `StateMachineThread.cpp`. Similarly, if there is a main class handling vision processing, name the file `VisionThread.py`. Then, create a class with the same name as the file and inherit the `AutonomyThread` interface and define the template type like shown below:
```
Comment (Byrdman32) [inline src/threads/README.md]: VisionThread.py ???

I thought that we were switching Autonomy from Python to C++.

**Review by Byrdman32 — CHANGES_REQUESTED** [review]

**Comment by Byrdman32** [comment]
Also, make sure to merge development now that the #89 has been merged.

File: src/threads/README.md (hunk 1 above)
Comment (ClayJay3) [reply src/threads/README.md]: fixed VisionThread.py

**Review by Byrdman32 — APPROVED** [review]
Approved

**Comment by Byrdman32** [comment]
@ClayJay3 Donovan says that it looks good and approves the PR.

**Comment by Byrdman32** [comment]
@ClayJay3 Jesse says that it looks good and approves the PR.

**Comment by ClayJay3** [comment]
Why don't they just approve it?

**Comment by jessedeuel** [comment]
I don't have the correct permissions. Eli is gonna fix it tomorrow.

**Comment by Byrdman32** [comment]
Because GitHub is being stupid and wasn't letting Jesse approve it. And Donovan didn't have access to his computer when he reviewed it and asked me to mark it for him.

File: src/threads/README.md [hunk 2]
```diff
@@ -1,38 +1,204 @@
...
-
-Here's a general guideline for organizing files within the **Threads** directory:
-
-- Create a separate file for each main class that requires its own dedicated thread.
-- Use descriptive names for the files that reflect the purpose or functionality of each class.
-- Inherit the `AutonomyThread` abstract class in the interfaces folder, to ensure the class you want to thread will integrate with the rest of the codebase.
-- Ensure that the files are properly documented with comments explaining the class's role, its interaction with other modules, and any relevant details.
-- Implement the necessary thread management mechanisms within each class, adhering to best practices for multithreading.
-
-For example, if there is a main class responsible for managing the states of the rover, create a file named `StateMachineThread.cpp`. Similarly, if there is a main class handling vision processing, name the file as `VisionThread.py`.
```
Comment (wiidler) [inline src/threads/README.md:21]: should it be VisionThread.cpp? just want to make sure it doesn't confuse people since we switched to C++

**Review by wiidler — COMMENTED** [review]
Looks awesome, sorry for the delay in getting a review in

File: src/threads/README.md (hunk 2 above)
Comment (ClayJay3) [reply src/threads/README.md:21]: Yeah, I fixed it in a new commit, this is the old side.

**Review by wiidler — APPROVED** [review]

**Pull request merged**